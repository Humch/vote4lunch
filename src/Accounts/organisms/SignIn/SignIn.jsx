import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import PropTypes from 'prop-types';

import { Redirect } from 'react-router-dom';

import {
  Container,
  Row,
  Col,
  Button,
  Form,
} from 'reactstrap';

import { FormGroupPassword, FormGroupPseudo } from '../..';

import { loginUser, resetLoginFailure } from '../../actions';
import { onTopAlert, offTopAlert } from '../../../Core/actions';

import './SignIn.scss';

class SignIn extends Component {
  submitForm(e) {
    const {
      pseudo,
      password,
      loginUser: loginU,
    } = this.props;

    e.preventDefault();

    loginU(pseudo, password);
  }

  render() {
    const {
      authenticated,
      loginError,
      onTopAlert: onTA,
      offTopAlert: offTA,
      resetLoginFailure: resetLF,
    } = this.props;

    if (loginError) {
      onTA('danger', 'votre pseudo ou votre mot de passe est erronÃ©');
      resetLF();
      setTimeout(() => {
        offTA();
      }, 5000);
    }

    if (authenticated) {
      return (<Redirect to="/" />);
    }

    return (
      <Container fluid className="SignIn">
        <Row noGutters className="justify-content-center align-items-center h-100">
          <Col
            xs="12"
            md="4"
            className="bg-blue p-5 rounded"
          >
            <Form onSubmit={e => this.submitForm(e)}>
              <FormGroupPseudo
                text="Indique ton pseudo"
                forProp="pseudo"
                colorLabel="white"
                noCheck
              />
              <FormGroupPassword
                text="Indique ton mot de passe"
                forProp="password"
                colorLabel="white"
              />
              <div className="text-center mt-5">
                <Button color="success">Se connecter</Button>
              </div>
            </Form>
          </Col>

        </Row>
      </Container>
    );
  }
}

SignIn.propTypes = {
  loginUser: PropTypes.func.isRequired,
  onTopAlert: PropTypes.func.isRequired,
  offTopAlert: PropTypes.func.isRequired,
  resetLoginFailure: PropTypes.func.isRequired,
  pseudo: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  authenticated: PropTypes.bool.isRequired,
  loginError: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool
  ]),
};

const mstp = ({ userData }) => ({
  pseudo: userData.pseudo,
  password: userData.password,
  authenticated: userData.authenticated,
  loginError: userData.loginError,
});

const mdtp = dispatch => bindActionCreators({
  loginUser,
  resetLoginFailure,
  onTopAlert,
  offTopAlert,
}, dispatch);

export default connect(mstp, mdtp)(SignIn);
