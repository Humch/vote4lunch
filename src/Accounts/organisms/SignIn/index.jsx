import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { func } from 'prop-types';

import { Redirect } from 'react-router-dom';

import { userDataTypes } from '../../types';

import { FormGroupSignIn } from '../..';

import { resetLoginFailure } from '../../actions';
import { onTopAlert, offTopAlert } from '../../../Core/actions';

import './SignIn.scss';

class SignIn extends Component {
  displayLoginError = () => {
    const {
      onTopAlert: onTA,
      offTopAlert: offTA,
      resetLoginFailure: resetLF,
    } = this.props;

    onTA('danger', 'votre pseudo ou votre mot de passe est erronÃ©');
    resetLF();
    setTimeout(() => {
      offTA();
    }, 5000);
  }

  render() {
    const { userData } = this.props;

    if (userData.loginError) {
      this.displayLoginError();
    }

    if (userData.authenticated) {
      return (<Redirect to="/" />);
    }

    return (
      <div className="SignIn container-fluid">
        <div className="row no-gutters justify-content-center align-items-center h-100">
          <div
            xs="12"
            md="4"
            className="col-12 col-md-4 bg-blue p-5 rounded"
          >
            <FormGroupSignIn />
          </div>
        </div>
      </div>
    );
  }
}

SignIn.propTypes = {
  onTopAlert: func.isRequired,
  offTopAlert: func.isRequired,
  resetLoginFailure: func.isRequired,
  userData: userDataTypes.isRequired,
};

const mstp = ({ userData }) => ({
  userData,
});

const mdtp = dispatch => bindActionCreators({
  resetLoginFailure,
  onTopAlert,
  offTopAlert,
}, dispatch);

export default connect(mstp, mdtp)(SignIn);
