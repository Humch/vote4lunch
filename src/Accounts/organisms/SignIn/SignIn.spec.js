/* eslint-disable no-undef */
import React from 'react';
import reduxThunk from 'redux-thunk';
import { Provider } from 'react-redux';

import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { BrowserRouter as Router } from 'react-router-dom';

import SignIn from './SignIn';

describe('SignIn Snapshot', () => {
  const initialState = {
    userData: {
      pseudo: '',
      password: '',
      passwordRepeater: '',
      authenticated: false,
      loginError: false,
    },
    pseudoChecker: {
      payload: {
        exist: false,
      },
      loading: false,
    },
  };
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('capturing Snapshot of SignIn', () => {
    const renderedValue = renderer.create(
      <Provider store={store}>
        <SignIn />
      </Provider>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe('SignIn Snapshot with loginError true', () => {
  const initialState = {
    userData: {
      pseudo: '',
      password: '',
      passwordRepeater: '',
      authenticated: false,
      loginError: true,
    },
    pseudoChecker: {
      payload: {
        exist: false,
      },
      loading: false,
    },
  };
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('capturing Snapshot of SignIn with loginError true', () => {
    const renderedValue = renderer.create(
      <Provider store={store}>
        <SignIn />
      </Provider>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe('SignIn (Shallow + passing the {store} directly)', () => {
  const initialState = {
    userData: {
      pseudo: '',
      password: '',
      passwordRepeater: '',
      authenticated: false,
      loginError: false,
    },
    pseudoChecker: {
      payload: {
        exist: false,
      },
      loading: false,
    },
  };
  const mockStore = configureStore();
  let store;
  let container;

  beforeEach(() => {
    store = mockStore(initialState);
    container = shallow(<SignIn store={store} />);
  });

  it('render the connected(SMART) component', () => {
    expect(container.length).toEqual(1);
  });
});

describe('SignIn - REACT-REDUX (Mount + wrapping in <Provider>)', () => {
  const initialState = {
    userData: {
      pseudo: 'bob',
      password: 'f4k3password',
      passwordRepeater: '',
      authenticated: false,
      loginError: false,
    },
    pseudoChecker: {
      payload: {
        exist: false,
      },
      loading: false,
    },
  };
  const middlewares = [reduxThunk];
  const mockStore = configureStore(middlewares);
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();

    wrapper = mount(
      <Provider store={store}>
        <SignIn />
      </Provider>
    );
  });

  it('render the connected(SMART) component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('check if form is submit', () => {
    const signUpForm = wrapper.find('form');
    signUpForm.simulate('submit', { preventDefault: jest.fn() });
    const action = store.getActions();

    expect(action[0].type).toBe('LOGIN_USER_BEGIN');
  });
});

describe('SignIn - REACT-REDUX (Mount + wrapping in <Provider>) - test Redirect', () => {
  const initialState = {
    userData: {
      pseudo: 'bob',
      password: 'f4k3password',
      passwordRepeater: '',
      authenticated: true,
      loginError: false,
    },
    pseudoChecker: {
      payload: {
        exist: false,
      },
      loading: false,
    },
  };
  const middlewares = [reduxThunk];
  const mockStore = configureStore(middlewares);
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();

    wrapper = mount(
      <Router>
        <Provider store={store}>
          <SignIn />
        </Provider>
      </Router>
    );
  });

  it('render the connected(SMART) component', () => {
    const redirect = wrapper.find('Redirect');

    expect(redirect.length).toEqual(1);
    expect(wrapper.length).toEqual(1);
  });
});
