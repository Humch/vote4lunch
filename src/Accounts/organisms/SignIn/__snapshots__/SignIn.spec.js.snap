// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Snapshot capturing Snapshot of SignIn 1`] = `
<div
  className="SignIn container-fluid"
>
  <div
    className="row no-gutters justify-content-center align-items-center h-100"
  >
    <div
      className="col-12 col-md-4 bg-blue p-5 rounded"
      md="4"
      xs="12"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className=" text-white"
            htmlFor="pseudo"
          >
            Indique ton pseudo
          </label>
          <div
            className="input-group"
          >
            <input
              className="form-control "
              id="pseudo"
              name="pseudo"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
            
            <div
              className="invalid-feedback"
            >
              Ce pseudo existe déjà
            </div>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=" text-white"
            htmlFor="password"
          >
            Indique ton mot de passe
          </label>
          <div
            className="input-group"
          >
            <input
              className="form-control "
              id="password"
              name="password"
              onChange={[Function]}
              required={true}
              type="password"
              value=""
            />
            
            
          </div>
        </div>
        <div
          className="text-center mt-5"
        >
          <button
            className="btn btn-success"
            type="submit"
          >
            Se connecter
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`SignIn Snapshot with loginError true capturing Snapshot of SignIn with loginError true 1`] = `
<div
  className="SignIn container-fluid"
>
  <div
    className="row no-gutters justify-content-center align-items-center h-100"
  >
    <div
      className="col-12 col-md-4 bg-blue p-5 rounded"
      md="4"
      xs="12"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className=" text-white"
            htmlFor="pseudo"
          >
            Indique ton pseudo
          </label>
          <div
            className="input-group"
          >
            <input
              className="form-control "
              id="pseudo"
              name="pseudo"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
            
            <div
              className="invalid-feedback"
            >
              Ce pseudo existe déjà
            </div>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=" text-white"
            htmlFor="password"
          >
            Indique ton mot de passe
          </label>
          <div
            className="input-group"
          >
            <input
              className="form-control "
              id="password"
              name="password"
              onChange={[Function]}
              required={true}
              type="password"
              value=""
            />
            
            
          </div>
        </div>
        <div
          className="text-center mt-5"
        >
          <button
            className="btn btn-success"
            type="submit"
          >
            Se connecter
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
